File: app/src/main/java/space/pal/sig/view/launches/LaunchActivity.java
package space.pal.sig.view.launches;

import android.app.Dialog;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.os.CountDownTimer;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.LinearLayout;

import androidx.annotation.NonNull;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.widget.AppCompatImageButton;
import androidx.appcompat.widget.AppCompatImageView;
import androidx.appcompat.widget.AppCompatTextView;
import androidx.appcompat.widget.Toolbar;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.appbar.AppBarLayout;
import com.google.android.material.appbar.CollapsingToolbarLayout;
import com.squareup.picasso.Picasso;

import java.util.Calendar;
import java.util.List;
import java.util.Locale;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.OnClick;
import space.pal.sig.R;
import space.pal.sig.Space;
import space.pal.sig.model.Launch;
import space.pal.sig.repository.dto.LocationDto;
import space.pal.sig.repository.dto.MissionDto;
import space.pal.sig.repository.dto.RocketDto;
import space.pal.sig.view.SpaceBaseActivity;

/**
 * SpaceNow
 * Created by Catalin on 7/26/2020
 **/
public class LaunchActivity extends SpaceBaseActivity {

    public static final String LAUNCH_ID = "launch_id";

    @BindView(R.id.launch_rocket_image_cover)
    AppCompatImageView coverImage;
    @BindView(R.id.launch_title)
    AppCompatTextView title;
    @BindView(R.id.launch_favorite)
    AppCompatImageButton favorite;
    @BindView(R.id.launch_cover_layout)
    ConstraintLayout coverLayout;
    @BindView(R.id.launch_appbar)
    AppBarLayout appBarLayout;
    @BindView(R.id.launch_collapsing_toolbar)
    CollapsingToolbarLayout collapsingToolbarLayout;
    @BindView(R.id.launch_rocket_image)
    AppCompatImageView rocketImage;
    @BindView(R.id.days)
    AppCompatTextView days;
    @BindView(R.id.hours)
    AppCompatTextView hours;
    @BindView(R.id.minutes)
    AppCompatTextView minutes;
    @BindView(R.id.seconds)
    AppCompatTextView seconds;
    @BindView(R.id.launch_status)
    AppCompatTextView launchStatus;
    @BindView(R.id.divider2)
    View divider2;
    @BindView(R.id.launch_clock)
    LinearLayout clockLayout;
    @BindView(R.id.launch_missions_label)
    AppCompatTextView missionsLabel;
    @BindView(R.id.launch_missions_recycler)
    RecyclerView missionsRecycler;
    @BindView(R.id.launch_date)
    AppCompatTextView launchDate;
    @BindView(R.id.launch_location_name)
    AppCompatTextView locationName;
    @BindView(R.id.launch_rocket_name)
    AppCompatTextView rocketName;
    @BindView(R.id.launch_rocket_configuration)
    AppCompatTextView rocketConfiguration;
    @BindView(R.id.launch_rocket_family)
    AppCompatTextView rocketFamily;
    @Inject
    LaunchViewModelFactory factory;
    @BindView(R.id.launch_toolbar)
    Toolbar toolbar;
    private LaunchViewModel launchViewModel;
    private CountDownTimer countDownTimer;
    private LaunchMissionsAdapter launchMissionsAdapter;
    private Launch launch;
    private Menu menu;
    private long launchId;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_launch);
        Space.getApplicationComponent().inject(this);
        int titleBarHeight = getStatusBarHeight();
        CollapsingToolbarLayout.LayoutParams params =
                (CollapsingToolbarLayout.LayoutParams) coverLayout.getLayoutParams();
        params.bottomMargin = -titleBarHeight;
        coverLayout.setLayoutParams(params);
        setSupportActionBar(toolbar);
        setTitle("");
        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setHomeAsUpIndicator(R.drawable.ic_baseline_close_24);
            actionBar.setDisplayHomeAsUpEnabled(true);
        }
        launchId = getIntent().getLongExtra(LAUNCH_ID, 0);
        launchViewModel = new ViewModelProvider(this, factory).get(LaunchViewModel.class);
        launchViewModel.getLaunch().observe(this, this::consumeLaunch);
        launchViewModel.loadLaunch(launchId);
        launchMissionsAdapter = new LaunchMissionsAdapter();
        missionsRecycler.setLayoutManager(new LinearLayoutManager(this,
                LinearLayoutManager.HORIZONTAL, false));
        missionsRecycler.setAdapter(launchMissionsAdapter);
    }

    @Override
    public boolean onSupportNavigateUp() {
        onBackPressed();
        return super.onSupportNavigateUp();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        this.menu = menu;
        boolean hasNotifications = launchViewModel
                .hasNotifications(launchId);
        int menuId = hasNotifications ? R.menu.menu_launch_notification : R.menu.menu_launch;
        getMenuInflater().inflate(menuId, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        switch (item.getItemId()) {
            case R.id.watch:
                if (launch.getVideoUrls() != null) {
                    final String[] videoUrls = launch.getVideoUrls().split(",");
                    Dialog dialog = new AlertDialog
                            .Builder(this, R.style.AppTheme_Dialog)
                            .setTitle(R.string.watch)
                            .setItems(videoUrls, (dialogInterface, i) -> {
                                Intent intent = new Intent(Intent.ACTION_VIEW);
                                intent.setData(Uri.parse(videoUrls[i]));
                                if (intent.resolveActivity(getPackageManager()) != null) {
                                    startActivity(intent);
                                }
                                dialogInterface.dismiss();
                            })
                            .create();
                    dialog.show();
                }
                break;
            case R.id.notification:
                final String[] options = new String[]{"On", "Off"};
                Dialog dialog = new AlertDialog
                        .Builder(this, R.style.AppTheme_Dialog)
                        .setTitle(R.string.notification)
                        .setItems(options, (dialogInterface, i) -> {
                            launchViewModel.toggleNotifications(i == 0, launchId);
                            dialogInterface.dismiss();
                            supportInvalidateOptionsMenu();
                            onCreateOptionsMenu(menu);
                        })
                        .create();
                dialog.show();
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    @OnClick(R.id.view_location)
    public void viewLocation() {
        LocationDto locationDto = launch.getLocation();
        if (locationDto == null) return;
        double latitude = Double.parseDouble(locationDto.getPads().get(0).getLatitude());
        double longitude = Double.parseDouble(locationDto.getPads().get(0).getLongitude());
        String uri = String.format(Locale.ENGLISH, "geo:%f,%f", latitude, longitude);
        Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse(uri));
        startActivity(intent);
    }

    @OnClick(R.id.launch_favorite)
    public void toggleFavorite() {
        launchViewModel.toggleFavorite(launch);
    }

    @Override
    protected void onDestroy() {
        if (countDownTimer != null) {
            countDownTimer.cancel();
        }
        super.onDestroy();
    }

    private void consumeLaunch(Launch launch) {
        this.launch = launch;
        RocketDto rocketDto = launch.getRocket();
        LocationDto locationDto = launch.getLocation();
        List<MissionDto> missionDtoList = launch.getMissions();
        if (rocketDto == null) return;
        if (launch.getFavorite()) {
            favorite.setImageResource(R.drawable.ic_baseline_favorite_24);
        } else {
            favorite.setImageResource(R.drawable.ic_baseline_favorite_border_24);
        }
        title.setText(launch.getName());
        if (missionDtoList == null || missionDtoList.size() == 0) {
            missionsLabel.setVisibility(View.GONE);
            missionsRecycler.setVisibility(View.GONE);
        } else {
            launchMissionsAdapter.submitList(missionDtoList);
            missionsLabel.setVisibility(View.VISIBLE);
            missionsRecycler.setVisibility(View.VISIBLE);
        }
        launchDate.setText(launch.getDate());
        if (locationDto != null && locationDto.getPads().size() > 0) {
            locationName.setText(locationDto.getPads().get(0).getName());
        }
        Picasso.get().load(rocketDto.getImageURL())
                .centerCrop()
                .fit()
                .into(coverImage);
        Picasso.get().load(rocketDto.getImageURL())
                .centerCrop()
                .fit()
                .into(rocketImage);
        rocketName.setText(rocketDto.getName());
        rocketConfiguration.setText(String.format("Configuration: %s", rocketDto.getConfiguration()));
        if (rocketDto.getFamily() != null) {
            rocketFamily.setText(String.format("Family: %s", rocketDto.getFamily().getName()));
        }
        switch (launch.getStatus()) {
            case GREEN:
                long now = Calendar.getInstance().getTimeInMillis();
                long countDown = launch.getTimestamp() - now;
                countDownTimer = new CountDownTimer(countDown, 1000) {
                    @Override
                    public void onTick(long l) {
                        long secondsInMilli = 1000;
                        long minutesInMilli = secondsInMilli * 60;
                        long hoursInMilli = minutesInMilli * 60;
                        long daysInMilli = hoursInMilli * 24;
                        long d = l / daysInMilli;
                        l = l % daysInMilli;
                        long h = l / hoursInMilli;
                        l = l % hoursInMilli;
                        long m = l / minutesInMilli;
                        l = l % minutesInMilli;
                        long s = l / secondsInMilli;
                        String dString = formatNumber(d);
                        String hString = formatNumber(h);
                        String mString = formatNumber(m);
                        String sString = formatNumber(s);
                        if (days == null || hours == null || minutes == null || seconds == null)
                            return;
                        days.setText(dString);
                        hours.setText(hString);
                        minutes.setText(mString);
                        seconds.setText(sString);
                    }

                    @Override
                    public void onFinish() {

                    }
                };
                countDownTimer.start();
                clockLayout.setVisibility(View.VISIBLE);
                divider2.setVisibility(View.VISIBLE);
                launchStatus.setBackgroundResource(R.drawable.launch_status_success);
                break;
            case SUCCESS:
                launchStatus.setBackgroundResource(R.drawable.launch_status_success);
                clockLayout.setVisibility(View.GONE);
                divider2.setVisibility(View.GONE);
                break;
            case RED:
            case FAILED:
                launchStatus.setBackgroundResource(R.drawable.launch_status_failure);
                clockLayout.setVisibility(View.GONE);
                divider2.setVisibility(View.GONE);
                break;
        }
        launchStatus.setText(launch.getStatus().getName());
    }

    private String formatNumber(long number) {
        String numberStr;
        if (number < 10) {
            numberStr = String.format("0%s", number);
        } else {
            numberStr = String.valueOf(number);
        }
        return numberStr;
    }

    private int getStatusBarHeight() {
        int result = 0;
        int resourceId = getResources()
                .getIdentifier("status_bar_height", "dimen", "android");
        if (resourceId > 0) {
            result = getResources().getDimensionPixelSize(resourceId);
        }
        return result;
    }
}

File: app/src/main/java/space/pal/sig/view/launch/LaunchActivity.kt
package space.pal.sig.view.launch

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.os.CountDownTimer
import android.widget.Toast
import androidx.core.view.isVisible
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.SupportMapFragment
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.MarkerOptions
import com.google.android.material.appbar.CollapsingToolbarLayout
import com.squareup.picasso.Picasso
import org.koin.android.ext.android.inject
import space.pal.sig.R
import space.pal.sig.databinding.ActivityLaunchBinding
import space.pal.sig.model.entity.CrewMember
import space.pal.sig.model.entity.LaunchPad
import space.pal.sig.model.entity.LaunchWithData
import space.pal.sig.util.ActivityExtensions.getStatusBarHeight
import space.pal.sig.util.DateExtensions.displayDatetime
import space.pal.sig.view.BaseActivity
import space.pal.sig.view.extra.WebActivity
import java.util.*

class LaunchActivity : BaseActivity(), CrewClickListener {

    companion object {
        const val LAUNCH_ID = "launch_id"
    }

    private lateinit var binding: ActivityLaunchBinding
    private val viewModel: LaunchViewModel by inject()
    private val adapter = CrewAdapter(this)
    private var countDownTimer: CountDownTimer? = null
    private var launchId: String? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = DataBindingUtil.setContentView(this, R.layout.activity_launch)
        launchId = intent.getStringExtra(LAUNCH_ID)
        viewModel.submitLaunchId(launchId)
        viewModel.getLaunch()
                .observe(this) { showLaunch(it) }
        viewModel.getCurrentRocketImage()
                .observe(this) { showRocketImage(it) }
        viewModel.getCurrentLaunchImage()
                .observe(this) { showLaunchImage(it) }
        viewModel.getCrew()
                .observe(this) { showCrew(it) }
        setupCoverAndToolbar()
        binding.launchNotificationsBtn
                .setOnClickListener { onNotificationBtnClick() }
        binding.launchWatchBtn
                .setOnClickListener { onWatchBtnClick() }
    }

    override fun onDestroy() {
        countDownTimer?.cancel()
        super.onDestroy()
    }

    override fun onSupportNavigateUp(): Boolean {
        onBackPressed()
        return super.onSupportNavigateUp()
    }

    override fun onClick(crewMember: CrewMember) {
        val intent = Intent(this, WebActivity::class.java)
        intent.putExtra(WebActivity.URL, crewMember.wikipedia)
        intent.putExtra(WebActivity.TITLE, crewMember.name)
        startActivity(intent)
    }

    private fun showRocketImage(image: String) {
        Picasso.get().load(image)
                .fit()
                .centerCrop()
                .into(binding.launchRocketImage)
    }

    private fun showLaunchImage(image: String) {
        Picasso.get().load(image)
                .fit()
                .centerCrop()
                .into(binding.launchRocketImageCover)
    }

    private fun showCrew(crew: List<CrewMember>?) {
        binding.launchCrewRecycler.layoutManager = LinearLayoutManager(
                this,
                LinearLayoutManager.HORIZONTAL,
                false)
        binding.launchCrewRecycler.adapter = adapter
        crew?.let {
            adapter.submit(it)
            binding.launchCrewRecycler.isVisible = it.isNotEmpty()
        }
    }

    private fun showLaunch(launchWithData: LaunchWithData) {
        val launch = launchWithData.launch
        val rocket = launchWithData.rocket
        val launchpad = launchWithData.launchpad
        val now = Calendar.getInstance().timeInMillis
        val countDown: Long = launch.dateUnix - now
        if (countDown > 0) {
            countDownTimer = object : CountDownTimer(countDown, 1000) {
                override fun onTick(l: Long) {
                    var millis = l
                    val secondsInMilli: Long = 1000
                    val minutesInMilli = secondsInMilli * 60
                    val hoursInMilli = minutesInMilli * 60
                    val daysInMilli = hoursInMilli * 24
                    val d = millis / daysInMilli
                    millis %= daysInMilli
                    val h = millis / hoursInMilli
                    millis %= hoursInMilli
                    val m = millis / minutesInMilli
                    millis %= minutesInMilli
                    val s = millis / secondsInMilli
                    val dString = formatNumber(d)
                    val hString = formatNumber(h)
                    val mString = formatNumber(m)
                    val sString = formatNumber(s)
                    binding.days.text = dString
                    binding.hours.text = hString
                    binding.minutes.text = mString
                    binding.seconds.text = sString
                }

                override fun onFinish() {}
            }
            countDownTimer?.start()
            binding.launchCountdownLayout.isVisible = true
            binding.divider2.isVisible = true
            binding.launchStatusTv.setBackgroundResource(R.drawable.launch_status_success)
            binding.launchStatusTv.text = getString(R.string.launch_available)
            binding.launchNotificationsBtn.isVisible = true
        } else {
            binding.launchCountdownLayout.isVisible = false
            binding.divider2.isVisible = false
            binding.launchStatusTv.text = getString(R.string.launch_unknown)
            binding.launchNotificationsBtn.isVisible = false
            launch.success?.let {
                if (it) {
                    binding.launchStatusTv.setBackgroundResource(R.drawable.launch_status_success)
                    binding.launchStatusTv.text = getString(R.string.launch_success)
                } else {
                    binding.launchStatusTv.setBackgroundResource(R.drawable.launch_status_failure)
                    binding.launchStatusTv.text = getString(R.string.launch_failure)
                }
            }
        }
        binding.launchTitle.text = launch.name
        binding.launchDate.text = getString(R.string.launch_date_label,
                Date(launch.dateUnix).displayDatetime())
        launchpad?.let {
            binding.launchLocationTv.text = it.fullName
            binding.launchLocationCard.isVisible = true
            setupMap(it)
        }
        launch.details?.let {
            binding.launchDetailsTv.text = it
            binding.launchDetailsCard.isVisible = true
        }
        launch.links.webcast?.let {
            binding.launchWatchBtn.tag = it
            binding.launchWatchBtn.isVisible = true
        }
        rocket?.let {
            val name = "${it.name} | ${it.company}"
            binding.launchRocketName.text = name
            binding.launchRocketDescription.text = it.description
            binding.launchRocketCard.isVisible = true
        }
        handleNotificationBtn(launch.id)
    }

    private fun setupMap(launchpad: LaunchPad) {
        val fragment: SupportMapFragment = supportFragmentManager
                .findFragmentById(R.id.launch_location_map) as SupportMapFragment
        fragment.getMapAsync { googleMap ->
            googleMap.mapType = GoogleMap.MAP_TYPE_SATELLITE
            googleMap.uiSettings.isZoomGesturesEnabled = false
            googleMap.uiSettings.isScrollGesturesEnabled = false
            if (launchpad.latitude != null && launchpad.longitude != null) {
                val latLng = LatLng(launchpad.latitude, launchpad.longitude)
                googleMap.addMarker(
                        MarkerOptions().position(latLng).title(launchpad.name)
                )
                val cameraUpdate = CameraUpdateFactory.newLatLngZoom(latLng, 7.0f)
                googleMap.animateCamera(cameraUpdate)
            }
        }
    }

    private fun handleNotificationBtn(launchId: String?) {
        val hasNotifications: Boolean = viewModel.hasNotifications(launchId)
        binding.launchNotificationsBtn.tag = hasNotifications
        if (hasNotifications) {
            binding.launchNotificationsBtn
                    .setImageResource(R.drawable.ic_baseline_notifications_active_24)
        } else {
            binding.launchNotificationsBtn
                    .setImageResource(R.drawable.ic_baseline_notifications_24)
        }
    }

    private fun onNotificationBtnClick() {
        val isOn = !(binding.launchNotificationsBtn.tag as Boolean)
        binding.launchNotificationsBtn.tag = isOn
        viewModel.toggleNotifications(isOn, launchId)
        handleNotificationBtn(launchId)
        val message = when (isOn) {
            true -> getString(R.string.launch_notification_on)
            else -> getString(R.string.launch_notification_off)
        }
        Toast.makeText(
                this,
                message,
                Toast.LENGTH_SHORT
        ).show()
    }

    private fun onWatchBtnClick() {
        val intent = Intent(Intent.ACTION_VIEW)
        intent.data = Uri.parse(binding.launchWatchBtn.tag as String)
        startActivity(intent)
    }

    private fun setupCoverAndToolbar() {
        val titleBarHeight: Int = getStatusBarHeight()
        val params = binding.launchCoverLayout.layoutParams as CollapsingToolbarLayout.LayoutParams
        params.bottomMargin = -titleBarHeight
        binding.launchCoverLayout.layoutParams = params
        title = null
        setSupportActionBar(binding.launchToolbar)
        supportActionBar?.let {
            it.setDisplayHomeAsUpEnabled(true)
            it.setHomeAsUpIndicator(R.drawable.ic_baseline_close_24)
        }
    }

    private fun formatNumber(number: Long): String {
        return if (number < 10) {
            String.format("0%s", number)
        } else {
            number.toString()
        }
    }
}